internal/api/api.go

```diff
@@ -0,0 +1,203 @@
+package api
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"net"
+
+	grpczap "github.com/grpc-ecosystem/go-grpc-middleware/v2/interceptors/logging"
+	"github.com/sierrasoftworks/humane-errors-go"
+	bladeapiv1alpha1 "github.com/uptime-industries/compute-blade-agent/api/bladeapi/v1alpha1"
+	agent2 "github.com/uptime-industries/compute-blade-agent/pkg/agent"
+	"github.com/uptime-industries/compute-blade-agent/pkg/events"
+	"github.com/uptime-industries/compute-blade-agent/pkg/log"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/status"
+	"google.golang.org/protobuf/types/known/emptypb"
+)
+
+type ListenMode string
+
+const (
+	ModeTcp  ListenMode = "tcp"
+	ModeUnix ListenMode = "unix"
+)
+
+func ListenModeFromString(s string) (ListenMode, humane.Error) {
+	switch s {
+	case string(ModeTcp):
+		return ModeTcp, nil
+	case string(ModeUnix):
+		return ModeUnix, nil
+	default:
+		return "", humane.New("invalid listen mode",
+			"ensure you are passing a valid listen mode to the grpc server",
+			"valid modes are: [tcp, unix]",
+		)
+	}
+}
+
+func (l ListenMode) String() string {
+	return string(l)
+}
+
+// AgentGrpcService represents a gRPC server implementation for managing compute blade agents.
+// It embeds UnimplementedBladeAgentServiceServer for forward compatibility and integrates ComputeBladeAgent logic.
+// The type allows for serving gRPC requests and gracefully shutting down the server.
+type AgentGrpcService struct {
+	bladeapiv1alpha1.UnimplementedBladeAgentServiceServer
+	agent         agent2.ComputeBladeAgent
+	server        *grpc.Server
+	authenticated bool
+	listenAddr    string
+	listenMode    ListenMode
+}
+
+// NewGrpcApiServer creates a new gRPC service
+func NewGrpcApiServer(ctx context.Context, options ...GrpcApiServiceOption) *AgentGrpcService {
+	service := &AgentGrpcService{}
+
+	for _, option := range options {
+		option(service)
+	}
+
+	grpcOpts := make([]grpc.ServerOption, 0)
+
+	// If we run our gRPC Server TLS with authentication enabled
+	if service.listenMode == ModeTcp && service.authenticated {
+		// Load server's certificate and private key
+		cert, certPool, err := EnsureServerCertificate(ctx)
+		if err != nil {
+			log.FromContext(ctx).Fatal("failed to load server key pair",
+				zap.Error(err),
+				zap.Strings("advice", err.Advice()),
+			)
+		}
+
+		// Create the TLS config that enforces mTLS for client authentication
+		tlsConfig := &tls.Config{
+			Certificates:       []tls.Certificate{cert},
+			ClientAuth:         tls.VerifyClientCertIfGiven, //tls.RequireAndVerifyClientCert, // FIXME
+			ClientCAs:          certPool,
+			InsecureSkipVerify: true, // FIXME
```

Comment:
I'd expose this as CLI flags

Reply:
AOh wow - good catch!
This is a left-over from debugging. I gotta remove this again. `InsecureSkipVerify` should most definitely be set to `false` and  `ClientAuth` should be set to `tls.RequireAndVerifyClientCert`